lib/segment/src/index/sparse_index/sparse_vector_index.rs

```diff
@@ -184,91 +134,177 @@ impl<TInvertedIndex: InvertedIndex> SparseVectorIndex<TInvertedIndex> {
+                        prefiltered_points.as_ref().unwrap().iter().copied()
+                    }
+                };
+                Ok(raw_scorer.peek_top_iter(&mut filtered_points, top))
+            }
+            None => Ok(raw_scorer.peek_top_all(top)),
         }
-        Ok(results)
     }
-}
 
-impl<TInvertedIndex: InvertedIndex> VectorIndex for SparseVectorIndex<TInvertedIndex> {
-    fn search(
+    fn search_sparse(
         &self,
-        vectors: &[&QueryVector],
+        sparse_vector: &SparseVector,
         filter: Option<&Filter>,
         top: usize,
-        _params: Option<&SearchParams>, // unused for sparse search
         is_stopped: &AtomicBool,
-    ) -> OperationResult<Vec<Vec<ScoredPointOffset>>> {
-        let id_tracker = self.id_tracker.borrow();
+    ) -> OperationResult<Vec<ScoredPointOffset>> {
         let vector_storage = self.vector_storage.borrow();
+        let id_tracker = self.id_tracker.borrow();
         let deleted_point_bitslice = id_tracker.deleted_point_bitslice();
         let deleted_vectors = vector_storage.deleted_vector_bitslice();
-        // filter for deleted points
+
         let not_deleted_condition = |idx: PointOffsetType| -> bool {
-            sparse_check_vector(idx, deleted_point_bitslice, deleted_vectors)
+            check_deleted_condition(idx, deleted_vectors, deleted_point_bitslice)
         };
+        let mut search_context = SearchContext::new(
+            sparse_vector.to_owned(),
+            top,
+            &self.inverted_index,
+            is_stopped,
+        );
+
         match filter {
             Some(filter) => {
                 let payload_index = self.payload_index.borrow();
-                let id_tracker = self.id_tracker.borrow();
-                let available_vector_count = vector_storage.available_vector_count();
-                let query_point_cardinality = payload_index.estimate_cardinality(filter);
-                let query_cardinality = adjust_to_available_vectors(
-                    query_point_cardinality,
-                    available_vector_count,
-                    id_tracker.available_point_count(),
-                );
+                let filter_context = payload_index.filter_context(filter);
+                let matches_filter_condition = |idx: PointOffsetType| -> bool {
+                    not_deleted_condition(idx) && filter_context.check(idx)
+                };
+                Ok(search_context.search(&matches_filter_condition))
+            }
+            None => Ok(search_context.search(&not_deleted_condition)),
+        }
+    }
 
-                // if cardinality is small - use plain search
-                if query_cardinality.max
-                    < self
+    pub fn search_query(
+        &self,
+        query_vector: &QueryVector,
+        filter: Option<&Filter>,
+        top: usize,
+        is_stopped: &AtomicBool,
+        prefiltered_points: &mut Option<Vec<PointOffsetType>>,
+        force_plain: bool,
```

Comment:
looks like this only used in benches. Why not call `search_plain` directly?

Reply:
Good idea! fixed